Disable FTZ/DAZ when compiling shared libraries by default.

jcranmer-intel opened the pull request:
This fixes https://github.com/llvm/llvm-project/issues/57589, and aligns Clang with the behavior of current versions of gcc. There is a new option, -mdaz-ftz, to control the linking of the file that sets FTZ/DAZ on startup, and this flag is on by default if -ffast-math is present and -shared isn't.

This also partially reverts fa7cd549d60 in that it disables the attempt to set the IR denormal-fp-math attribute based on whether or not -ffast-math is applied as it is insufficiently reliable.

jcranmer-intel committed:
Disable FTZ/DAZ when compiling shared libraries by default.

This fixes https://github.com/llvm/llvm-project/issues/57589, and aligns
Clang with the behavior of current versions of gcc. There is a new option,
-mdaz-ftz, to control the linking of the file that sets FTZ/DAZ on startup, and
this flag is on by default if -ffast-math is present and -shared isn't.

MaskRay commented on clang/include/clang/Driver/Options.td:
```diff
@@ -2554,6 +2554,11 @@ defm protect_parens : BoolFOption<"protect-parens",
+defm daz_ftz : SimpleMFlag<"daz-ftz",
+  "Globally set", "Do not globally set",
+  " the denormals-are-zero (DAZ) and flush-to-zero (FTZ) bits in the "
+  "floating-point control register on program startup.">;
```
The convention omits the trailing period.

jcranmer-intel committed:
Address review comments.

MaskRay commented on clang/test/Driver/linux-ld.c:
```diff
@@ -1569,6 +1569,40 @@
 // RUN:   | FileCheck --check-prefix=CHECK-NOCRTFASTMATH %s
+// Don't link crtfastmath.o with -shared
+// RUN: %clang --target=x86_64-unknown-linux -no-pie -### %s -ffast-math -shared \
+// RUN:        --gcc-toolchain="" \
```
You can drop `--gcc-toolchain=""`. Setting CMake `GCC_INSTALL_PREFIX` leads to a warning and the option will be removed.

-no-pie/-pie/-shared are mutual exclusive. The interaction is tested by other tests. For the purpose of fastmath, remove -no-pie when -shared is used.

andykaylor replied:
Do we have trouble with this test on systems where crtfastmath.o is not provided by the gcc toolchain? I don't think we provide our own version of this, do we?

jcranmer-intel replied:
The tests use a custom sysroot that provides a `crtfastmath.o` file, so it doesn't matter whether or not it's provided by the host toolchain.

jyknight commented:
I think there is a bit of a problematic interaction with getDenormalModeForType [here](https://github.com/llvm/llvm-project/blob/7a94acb2da5b20d12f13f3c5f4eb0f3f46e78e73/clang/lib/Driver/ToolChains/Linux.cpp#L838C8-L838C37). "-shared" is (should be) a flag used only for linking, but that function is calling isFastMathRuntimeAvailable to affect the default denormal math mode for _compilation_. That's not going to work.

Probably we should effectively revert the x86-linux parts of fa7cd549d604bfd8f9dce5d649a19720cbc39cca.

I wonder if, instead, we should just have `-ffast-math` always downgrade `-fdenormal-fp-math=ieee` to `-fdenormal-fp-math=preserve-sign`, under the rationale of "you asked for fast math, and preserve-sign mode might let the compiler generate faster code"?

arsenm commented:
> I wonder if, instead, we should just have `-ffast-math` always downgrade `-fdenormal-fp-math=ieee` to `-fdenormal-fp-math=preserve-sign`, under the rationale of "you asked for fast math, and preserve-sign mode might let the compiler generate faster code"?

This could also use denormal-fp-math=dynamic. It is not always safe to run preserve-sign code under IEEE settings

jyknight commented:
> It is not always safe to run preserve-sign code under IEEE settings

I can see that this is used in a bunch of optimization/constant-folding passes, but I don't have a feel for what the actual impact is:
1. Which value allows generating the "fastest" math code -- disregarding correctness? I'd assume that "dynamic" is least optimizable, "ieee" in the middle, and "preserve-sign" is likely to generate the "fastest" code?
2. What is the likely _actual_ impact of choosing the wrong value? Ie, what does "it is not always safe" mean? Is this wrong beyond what is usually expected from `-ffast-math` optimizations?

I note that we're frequently getting it wrong in the other direction today, as there are many users who link in code compiled _without_ `-ffast-math` (and thus, with `-fdenormal-fp-math=ieee`) into a binary compiled with `-ffast-math` -- which sets the fp control words to enable `MXCSR_DAZ | MXCSR_FTZ`. In addition, I think there's even more users who compile some TUs _with_ `-ffast-math`, and link that into a binary compiled without.

arsenm commented:
> * Which value allows generating the "fastest" math code -- disregarding correctness? I'd assume that "dynamic" is least optimizable, "ieee" in the middle, and "preserve-sign" is likely to generate the "fastest" code?

This depends on the target and operations. For some functions, on modern hardware, the preserve-sign code will be worse for AMDGPU. In other cases it's better 

> What is the likely actual impact of choosing the wrong value? Ie, what does "it is not always safe" mean? Is this wrong beyond what is usually expected from -ffast-math optimizations?

The main example is the conversions performed between fcmp with 0 and llvm.is.fpclass. You'll get different results depending on whether the input is implicitly flushed in fcmp vs. not in the is.fpclass

jcranmer-intel commented:
> I think there is a bit of a problematic interaction with getDenormalModeForType [here](https://github.com/llvm/llvm-project/blob/7a94acb2da5b20d12f13f3c5f4eb0f3f46e78e73/clang/lib/Driver/ToolChains/Linux.cpp#L838C8-L838C37). "-shared" is (should be) a flag used only for linking, but that function is calling isFastMathRuntimeAvailable to affect the default denormal math mode for _compilation_. That's not going to work.

The logic is already somewhat broken already: if `-ffast-math` isn't being added into the link line, or we're a shared library linked into a `-ffast-math` executable, we're not going to get the right value of denormal math mode. So it's a matter of which behavior is going to be the least likely to be incorrect in practice.

> I wonder if, instead, we should just have `-ffast-math` always downgrade `-fdenormal-fp-math=ieee` to `-fdenormal-fp-math=preserve-sign`, under the rationale of "you asked for fast math, and preserve-sign mode might let the compiler generate faster code"?

One of the issues is that we have optimizations that are going to make assumptions about how denormals work in operations (particularly around being able to lower `llvm.is.fpclass` to `fcmp`). So you risk misoptimization if the value is wrong. Arguably, the only safe value is dynamic (compiler doesn't know what it's set to).

(We're wrong in another direction, which is that we don't attempt to set the flag to non-ieee for anything other x86. I tried looking up how the equivalents of FTZ/DAZ worked on other architectures and came to with the conclusion that the whole thing is a bundle of sadness that should have never been invented.)

jyknight commented:
> You'll get different results depending on whether the input is implicitly flushed in fcmp vs. not in the is.fpclass

This sounds intuitively like the sort of semantics-breaking optimization which is expected from `-ffast-math`. If the only issues are things like getting a slightly-wrong answer from is.fpclass if you somehow pass a denormal to it, when you've disabled denormal handling...that just doesn't seem so bad.

> Arguably, the only safe value is dynamic (compiler doesn't know what it's set to).

Yes, we could change the default to "dynamic" for _all_ builds (except on platforms where FTZ/DAZ flags don't exist). Potentially doing so is justified regardless of `-ffast-math`, since users can of course use `fesetenv` or `_mm_setcsr` to enable these flags themselves. (Sidenote: "dynamic" isn't even [documented](https://clang.llvm.org/docs/UsersManual.html#cmdoption-fdenormal-fp-math)). 

> So it's a matter of which behavior is going to be the least likely to be incorrect in practice.

So, alternatively...we could just go with the simplest solution, and use "ieee" as the default even under `-ffast-math`.

jcranmer-intel commented:
> (Sidenote: "dynamic" isn't even [documented](https://clang.llvm.org/docs/UsersManual.html#cmdoption-fdenormal-fp-math)).

It's not a selectable enum of the Clang `-fdenormal-fp-math` flag, but it is one for the LLVM function attribute `denormal-fp-math`.

andykaylor commented:
I don't know anything about how non-x86 targets implement DAZ/FTZ, but for x86-based targets, I think trying to make any assumptions about the setting is bound to be wrong. In theory, it's part of the floating-point environment and shouldn't be modified during execution unless fenv-access is enabled.

In practical terms, I can't see any reason that you would ever want a shared library to be able to control this. I don't like the idea of adding an option to maintain that behavior, and I think we should make it a high priority to stop doing it by default when a shared library is compiled with fast-math enabled.

More generally, if you're going to be mixing fast-math with non-fast-math within a program, you don't want FTZ/DAZ to be enabled just because some parts of the program are using fast-math, so the current behavior of linking with crtfastmath.o is fairly reckless.

andykaylor commented:
> > (Sidenote: "dynamic" isn't even [documented](https://clang.llvm.org/docs/UsersManual.html#cmdoption-fdenormal-fp-math)).
> 
> It's not a selectable enum of the Clang `-fdenormal-fp-math` flag, but it is one for the LLVM function attribute `denormal-fp-math`.

This is also a bit of a problem. After inlining we can have instructions with fast-math disabled appearing in the middle of functions that were compiled with fast-math enabled. What happens to the "denormal-fp-math" attribute in such cases? I wanted to show this in compiler explorer, but "#pragma float_control()" doesn't update the "denormal-fp-math attribute.

https://godbolt.org/z/KEP3YEEe9

I may have mentioned a few times that I don't like function attributes controlling fast-math behaviors.

arsenm commented:
> I may have mentioned a few times that I don't like function attributes controlling fast-math behaviors.

It doesn't control it, it's informative. You just get undefined behavior if you end up calling mismatched mode functions.

It does control it in the AMDGPU entry point function case, since we are more or less directly programming the initial register state on kernel launch in the compiler

jyknight commented:
> So, alternatively...we could just go with the simplest solution, and use "ieee" as the default even under -ffast-math.

This is feeling like the best option to me, at this point. Easily implementable, and doesn't seem to make things significantly worse than they are today, since the denormal mode can't be controlled by compilation options -- so being wrong in both directions is going to happen, anyhow.

We'd simply delete [Linux::getDefaultDenormalModeForType](https://github.com/llvm/llvm-project/blob/d4c5acac99e83ffa12d2d720c9e502a181cbd7ea/clang/lib/Driver/ToolChains/Linux.cpp#L838), which thus causes `-ffast-math` to simply not change the compiler's understanding of denormals handling -- like is already the case on everything except x86/x86_64 Linux.

Other more-involved changes, such as exposing "-fdenormal-fp-math=dynamic" in Clang, and potentially making it the default, could be done later if desired. I'd suggest that this change doesn't need to block on any such discussions.

arsenm commented:
> > So, alternatively...we could just go with the simplest solution, and use "ieee" as the default even under -ffast-math.
> 

+1. There hasn't been a performance reason to use FTZ/DAZ since ~2011. Maybe there's still a power benefit? But in that case you could still explicitly request the flush separate from -ffast-math

andykaylor commented:
> > > So, alternatively...we could just go with the simplest solution, and use "ieee" as the default even under -ffast-math.
> 
> +1. **There hasn't been a performance reason to use FTZ/DAZ since ~2011.** Maybe there's still a power benefit? But in that case you could still explicitly request the flush separate from -ffast-math

I don't think this is correct. I know there have been improvements, but there are still cases where setting ftz can have a significant performance impact.

arsenm commented:
> Do you only set the register for kernel entries?

Yes, it's the pre-initialized state. Non kernels can't be arbitrarily invoked from the host 

>  Is the attribute ignored for other functions?

No, it's an informative attribute about that the mode is. The compiler isn't trying to fix up the modes for you. If you enter the function with the wrong mode, you get what you get.

jyknight commented:
> > > > So, alternatively...we could just go with the simplest solution, and use "ieee" as the default even under -ffast-math.
> > 
> > 
> > +1. **There hasn't been a performance reason to use FTZ/DAZ since ~2011.** Maybe there's still a power benefit? But in that case you could still explicitly request the flush separate from -ffast-math
> 
> I don't think this is correct. I know there have been improvements, but there are still cases where setting ftz can have a significant performance impact.

Just to be clear: I'm not proposing entirely eliminating the "link against crtfastmath.o" behavior, when linking a binary with `-ffast-math` (though, separately from _this_ review, that may be worth considering!). I only meant we should stop attempting to infer anything about `-fdenormal-fp-math` due to using `-ffast-math`. (as per the other paragraph in my last comment).

andykaylor commented:
> Just to be clear: I'm not proposing entirely eliminating the "link against crtfastmath.o" behavior, when linking a binary with `-ffast-math` (though, separately from _this_ review, that may be worth considering!). I only meant we should stop attempting to infer anything about `-fdenormal-fp-math` due to using `-ffast-math`. (as per the other paragraph in my last comment).

Are youe suggesting that we should continue linking against crtfastmath.o when it is available and fast-math is enabled but we should set the "denormal-fp-math" attributes to "ieee" or "dynamic"? I want to make sure I'm understanding you correctly.

Personally, I don't like linking with crtfastmath.o to get this behavior. I would prefer to insert code into whatever function(s), if any, we identify as a relevant entry point. That would require having the attribute set, at least for the entry function(s). Linking with crtfastmath.o depends on finding it, which I think makes it dependent on the GNU toolchain, and the static initializer is less visible to users debugging their program.

For x86-based CPU targets, we really shouldn't be making assumptions about the FTZ/DAZ state for any function where we aren't setting it, but if it's useful for other targets, we shouldn't take that away.

jyknight commented:
I'm proposing a simple change we can make _now_, in order to unblock this PR which at least gets rid of crtfastmath for shared libraries! I don't think this needs to be the end of the story; additional, more large-scale changes can be made afterwards...

> Are youe suggesting that we should continue linking against crtfastmath.o when it is available and fast-math is enabled

I'm suggesting no additional changes regarding when to link crtfastmath vs what's present in this PR.

> but we should set the "denormal-fp-math" attributes to "ieee" or "dynamic"? I want to make sure I'm understanding you correctly.

I'm suggesting that we modify Clang so that `-ffast-math` _doesn't affect_ `denormal-fp-math`, by (as I mentioned before) removing the overload [Linux::getDefaultDenormalModeForType](https://github.com/llvm/llvm-project/blob/d4c5acac99e83ffa12d2d720c9e502a181cbd7ea/clang/lib/Driver/ToolChains/Linux.cpp#L838). This makes Clang for Linux X86 and X86-64 work the same as every other OS/CPU combo.

Yes, this means we have the wrong mode, sometimes. But: that is not a _new_ problem -- we use the wrong mode all the time already. And ISTM the issues that arise are within the expected bounds of `-ffast-math` wrongness.

> For x86-based CPU targets, we really shouldn't be making assumptions about the FTZ/DAZ state for any function where we aren't setting it

If you'd make another PR proposing to expose `-fdenormal-fp-math=dynamic` to the Clang command-line, and making that the _default_ for most targets, that seems potentially interesting to discuss. I don't really know the impact, and I also don't think it needs to be concluded before making the changes for this PR.

andykaylor commented:
> I'm suggesting that we modify Clang so that `-ffast-math` _doesn't affect_ `denormal-fp-math`, by (as I mentioned before) removing the overload [Linux::getDefaultDenormalModeForType](https://github.com/llvm/llvm-project/blob/d4c5acac99e83ffa12d2d720c9e502a181cbd7ea/clang/lib/Driver/ToolChains/Linux.cpp#L838). This makes Clang for Linux X86 and X86-64 work the same as every other OS/CPU combo.

I don't see why the current denormal-fp-math setting behavior is a blocking issue for this change. Yes, compiling a shared library with ffast-math would lead to the "denormal-fp-math" attribute being set to a potentially misleading value, but that's already true for any file that is compiled with -ffast-math if the file containing the entry function isn't compiled with -ffast-math. I'd see fixing that as a separate issue.

I'd like to see this change land, but with the "-mdaz-ftz" option removed (because I don't think it's useful). That would fix the critical problem of fast-math infecting shared libraries with the ftz setting, and we could straighten out the other problems, which are relatively minor, afterwards.

jyknight commented:
> I don't see why the current denormal-fp-math setting behavior is a blocking issue for this change

Because this PR as-is causes us to start parsing the "-shared" flag for compilation actions in order to determine which denormal-fp-math setting to use. Users should not pass that to compile actions, and if they do, it should result in a `-Wunused-command-line-argument` diagnostic, NOT a behavior change.

jcranmer-intel commented:
> I'd like to see this change land, but with the "-mdaz-ftz" option removed (because I don't think it's useful). That would fix the critical problem of fast-math infecting shared libraries with the ftz setting, and we could straighten out the other problems, which are relatively minor, afterwards.

There is, without this change, no way to control whether or not `crtfastmath.o` is linked independent of all of the other fast-math options. The `-mdaz-ftz` option would at least add a flag to explicitly control the parameter (for the people who care), and we can then have discussions about different ways to effect setting DAZ/FTZ bits or what options imply `-mdaz-ftz` in future PRs. That alone makes it a worthy addition IMHO; the compatibility with gcc is another nice feature.

> I'm suggesting that we modify Clang so that -ffast-math doesn't affect denormal-fp-math, by (as I mentioned before) removing the overload [Linux::getDefaultDenormalModeForType](https://github.com/llvm/llvm-project/blob/d4c5acac99e83ffa12d2d720c9e502a181cbd7ea/clang/lib/Driver/ToolChains/Linux.cpp#L838). This makes Clang for Linux X86 and X86-64 work the same as every other OS/CPU combo.

That sounds to me ultimately like it should be a separate PR, although you're suggesting that should happen first?

andykaylor commented:
> > I'd like to see this change land, but with the "-mdaz-ftz" option removed (because I don't think it's useful). That would fix the critical problem of fast-math infecting shared libraries with the ftz setting, and we could straighten out the other problems, which are relatively minor, afterwards.
> 
> There is, without this change, no way to control whether or not `crtfastmath.o` is linked independent of all of the other fast-math options. The `-mdaz-ftz` option would at least add a flag to explicitly control the parameter (for the people who care), and we can then have discussions about different ways to effect setting DAZ/FTZ bits or what options imply `-mdaz-ftz` in future PRs. That alone makes it a worthy addition IMHO; the compatibility with gcc is another nice feature.

You can always link crtfastmath.o directly, of course. Ultimately, I don't think the compiler should ever be adding the crtfastmath.o file. I would prefer to insert code directly into the entry function as @arsenm indicated the AMDGPU backend does for kernels. That would then be controlled by the -fdenormal-fp-math option or something more explicitly linked to the entry function.

I don't want to add -mdaz-ftz because once we do we're kind of stuck with it. If you don't add it here, we'd continue the current behavior of linking with crtfastmath.o normally but we'd stop infecting shared libraries with it.

andykaylor commented:
> > I don't see why the current denormal-fp-math setting behavior is a blocking issue for this change
> 
> Because this PR as-is causes us to start parsing the "-shared" flag for compilation actions in order to determine which denormal-fp-math setting to use. Users should not pass that to compile actions, and if they do, it should result in a `-Wunused-command-line-argument` diagnostic, NOT a behavior change.

We're already parsing -nostartfiles and -nostdlib to determine the default setting. I don't see how looking at -shared makes it any worse.

I would agree that this whole model is broken. If I use -ffast-math to create a bunch of object files (including the one containing main()) and then I link without that option, crtfastmath.o doesn't get linked, even though we used "denorm-fp-math"="preserveSign" on every function.

I think we're in agreement that this needs to be fixed for x86-based targets. We just need to agree on how to do it. I think https://github.com/llvm/llvm-project/issues/57589 is a pretty egregious problem, and I'd like to see it fixed without delay, but given that it requires setting -ffast-math on your link line maybe it can wait for a proper fix.

andykaylor commented:
I just created https://github.com/llvm/llvm-project/issues/81204 to describe the linking versus compiling problem.

jcranmer-intel commented:
> > > I'd like to see this change land, but with the "-mdaz-ftz" option removed (because I don't think it's useful). That would fix the critical problem of fast-math infecting shared libraries with the ftz setting, and we could straighten out the other problems, which are relatively minor, afterwards.
> > 
> > 
> > There is, without this change, no way to control whether or not `crtfastmath.o` is linked independent of all of the other fast-math options. The `-mdaz-ftz` option would at least add a flag to explicitly control the parameter (for the people who care), and we can then have discussions about different ways to effect setting DAZ/FTZ bits or what options imply `-mdaz-ftz` in future PRs. That alone makes it a worthy addition IMHO; the compatibility with gcc is another nice feature.
> 
> You can always link crtfastmath.o directly, of course. Ultimately, I don't think the compiler should ever be adding the crtfastmath.o file. I would prefer to insert code directly into the entry function as @arsenm indicated the AMDGPU backend does for kernels. That would then be controlled by the -fdenormal-fp-math option or something more explicitly linked to the entry function.
> 
> I don't want to add -mdaz-ftz because once we do we're kind of stuck with it. If you don't add it here, we'd continue the current behavior of linking with crtfastmath.o normally but we'd stop infecting shared libraries with it.

I don't think it's unreasonable to switch the logic of `-mdaz-ftz` from linking a file with a global initializer that sets the flags to making it emit the entry-point call to such a function instead, it still largely follows the same logic to me. And having a command line flag makes it easier for users to access rather than manually linking in a file located who-knows-where in the toolchain (although I suspect anyone who cares hard enough would rather just write the calls to set FTZ/DAZ than track it down from the toolchain).

andykaylor commented:
> I don't think it's unreasonable to switch the logic of `-mdaz-ftz` from linking a file with a global initializer that sets the flags to making it emit the entry-point call to such a function instead, it still largely follows the same logic to me. And having a command line flag makes it easier for users to access rather than manually linking in a file located who-knows-where in the toolchain (although I suspect anyone who cares hard enough would rather just write the calls to set FTZ/DAZ than track it down from the toolchain).

I suppose there's also some portability value in supporting this gcc option. I guess I'd be OK with it.

andykaylor commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -117,6 +117,11 @@ C23 Feature Support
 -------------------------------------------------
 
+* Code compiled with ``-shared`` and ``-ffast-math`` will no longer enable
+  flush-to-zero floating-point mode by default. This decision can be overridden
```
You may want to revise this to clarify that this only refers to the use of -shared and -ffast-math on the command-line used to link the program.

andykaylor commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -117,6 +117,11 @@ C23 Feature Support
 Non-comprehensive list of changes in this release
 -------------------------------------------------
 
+* Code compiled with ``-shared`` and ``-ffast-math`` will no longer enable
```
```suggestion
* Code compiled with ``-shared`` and either ``-ffast-math`` or ``-funsafe-math-optimizations`` will no longer enable
```

jcranmer-intel committed:
Partially revert fa7cd549d60.

The guessing of whether or not FTZ/DAZ mode would be enabled is
insufficiently accurate.

jcranmer-intel commented:
I did a thorough investigation into how the `denormal-fp-math=preserve-sign,preserve-sign` attribute affects the resulting IR for all of the SPEC benchmarks (which actually do run into subnormals), and the basic summary I found is that the only differences I found were that the FTZ/DAZ mode prevented a few inferences that are still true in denormal-flushing: things like `x >= 0` implies `x >= -1`.

So overall, the practical effect of the `denormal-fp-math` attribute being set incorrectly doesn't appear to matter.

arsenm commented:
> So overall, the practical effect of the `denormal-fp-math` attribute being set incorrectly doesn't appear to matter.

It matters more for AMDGPU, where we need to care because some instructions just don't respect denormals. We legalize some operations differently depending on the mode

jcranmer-intel commented:
> It matters more for AMDGPU, where we need to care because some instructions just don't respect denormals. We legalize some operations differently depending on the mode

But the shared library stuff isn't an issue for AMDGPU, right?

arsenm commented:
> But the shared library stuff isn't an issue for AMDGPU, right?

No, we don't support shared library linking yet

jcranmer-intel commented:
Ping for review

arsenm commented on clang/lib/Driver/ToolChains/Linux.cpp:
```diff
@@ -842,25 +842,6 @@ void Linux::addProfileRTLibs(const llvm::opt::ArgList &Args,
-Linux::getDefaultDenormalModeForType(const llvm::opt::ArgList &DriverArgs,
-                                     const JobAction &JA,
-                                     const llvm::fltSemantics *FPType) const {
-  switch (getTriple().getArch()) {
```
Should shared libraries be using dynamic mode?

jcranmer-intel replied:
Probably. But we can't reliably tell from the individual file compile commands if we're going to be linked as a shared library or not, which is why I'm dropping this attempt to set it at all in the first place.

andykaylor replied:
So getDefaultDenormalModeForType() in ToolChain.h should be returning DenormalMode::Dynamic() rather than DenormalMode::getIEEE(), right?

jyknight replied:
No, we shouldn't. This was discussed before: https://github.com/llvm/llvm-project/pull/80475#issuecomment-1928170074 and surrounding comments.
But in short: doing so would regress normal code, only for the (arguable) benefit of code which is explicitly asks to play fast-and-loose with the rules via `-ffast-math`.

I think the change is good as-is.

andykaylor replied:
I didn't think that had been resolved, but I agree that this change is good as-is.

It's not just fast-math that can change ftz/daz though. Users can also modify MXCSR directly. In C, FENV_ACCESS needs to be enabled, but in practice I find that users would like to be able to do that without taking on the other penalties associated with enabling fenv access (such as generally diminished optimization). It would be nice if LLVM IR had a way to allow this.

jyknight reviewed (approved):
One final nit before you submit, though: please update the PR description and the release notes to also mention that x86 no longer modifies `-fdenormal-fp-math` based on `-ffast-math`.

jcranmer-intel committed:
Update release notes.

MaskRay commented on clang/lib/Driver/ToolChain.cpp:
```diff
@@ -1307,9 +1307,14 @@ void ToolChain::AddCCKextLibArgs(const ArgList &Args,
+  // Don't implicitly link in mode-changing libraries in a shared library, since
+  // this can have very deleterious effects. See the various links from
+  // https://github.com/llvm/llvm-project/issues/57589 for more information.
+  bool Default = !Args.hasArg(options::OPT_shared);
```
hasArgNoClaim

andykaylor commented:
> One final nit before you submit, though: please update the PR description and the release notes to also mention that x86 no longer modifies `-fdenormal-fp-math` based on `-ffast-math`.

The Clang User's Manual makes some contradictory statements about this. In one place it says, `-fno-fast-math` resets `-fdenormal-fp-math=<value>` to its target-dependent default, but just a few lines later it says "-fno-fast-math implies -fdenormal-fp-math=ieee" which isn't true for all targets. It says `-fno-unsafe-math-optimizations` implies '-fdenormal-fp-math=ieee` which, unfortunately, is currently be true.

The User's Manual also says that `-ffp-model=fast` sets `denormal_fp_math` to IEEE, which is currently true, but really shouldn't be.

I'm working on a PR to fix the fp-model handling, which I'll update after this PR lands. I can fix the other issues I've mentioned here in that patch. There are a few points that I am not quite clear on. I'm going to start a Discourse topic to summarize what we've agreed on here and discuss the things I'm not sure about.

jcranmer-intel committed:
Shift to argnoclaim to retain compiler warning.
